File: ParallelAndVectorized/Algorithms/MatrixMultiplicationThreads.py
import threading
import numpy as np

class MatrixMultiplicationThreads:
    def multiply(self, matrixA, matrixB, num_threads):
        size = len(matrixA)
        result = [[0.0 for _ in range(size)] for _ in range(size)]

        threads_per_row = max(1, size // num_threads)
        threads = []

        def multiply_rows(start_row, end_row):
            for row in range(start_row, end_row):
                self._multiply_row(row, matrixA, matrixB, result)

        for i in range(num_threads):
            start_row = i * threads_per_row
            end_row = min((i + 1) * threads_per_row, size)
            thread = threading.Thread(target=multiply_rows, args=(start_row, end_row))
            threads.append(thread)
            thread.start()

        for thread in threads:
            thread.join()

        return np.array(result)

    @staticmethod
    def _multiply_row(row, matrixA, matrixB, result):
        for j in range(len(matrixB[0])):
            for k in range(len(matrixA[0])):
                result[row][j] += matrixA[row][k] * matrixB[k][j]

File: ParalellAndVectorizedMatrices/src/main/java/org/example/MatrixMultiplicationThreads.java
package org.example;

public class MatrixMultiplicationThreads implements ParallelMatrixMultiplication {

    @Override
    public double[][] multiply(double[][] matrixA, double[][] matrixB, int numThreads) throws InterruptedException {
        int size = matrixA.length;
        double[][] result = new double[size][size];

        int threadsPerRow = Math.max(1, size / numThreads);
        Thread[] threads = new Thread[numThreads];

        for (int i = 0; i < numThreads; i++) {
            final int startRow = i * threadsPerRow;
            final int endRow = Math.min((i + 1) * threadsPerRow, size);

            threads[i] = new Thread(() -> {
                for (int row = startRow; row < endRow; row++) {
                    multiplyRow(row, matrixA, matrixB, result);
                }
            });
            threads[i].start();
        }

        for (int i = 0; i < numThreads; i++) {
            threads[i].join();
        }

        return result;
    }

    private static void multiplyRow(int row, double[][] matrixA, double[][] matrixB, double[][] result) {
        for (int j = 0; j < matrixB[0].length; j++) {
            for (int k = 0; k < matrixA[0].length; k++) {
                result[row][j] += matrixA[row][k] * matrixB[k][j];
            }
        }
    }

}
